components/board/genericlinux/gpio.go

```diff
@@ -22,7 +22,8 @@ type gpioPin struct {
 	line       *gpio.Line
 
 	// These values are mutable. Lock the mutex when interacting with them.
-	pwmRunning      bool
+	swPwmRunning    bool
+	hwPwm           *pwmDevice // Defined in hw_pwm.go, will be nil for pins that don't support it.
```

Comment:
[q] It does not seem from this code as if we're trying to support use of software PWM on a pin that supports hardware PWM. So why not have your `pwmDevice` struct simply satisfy the `GPIOPin` interface and be stored in `gpioMappings` directly rather than wrapping it inside another pin struct?

Reply:
Hardware PWM does not support very low frequencies: it definitely can't do 1 Hz, and might not be able to do other frequencies under 2 Hz (but I don't know because we only allow an integer frequency). Even if a pin has hardware PWM, we still need to be able to do software PWM at 1Hz.